File: spec/src/test/java/co/mscp/DockerUtil.java
package co.mscp;

import java.io.IOException;
import java.net.URL;
import java.util.Map;


public class DockerUtil {

    // --- INNER TYPES --- //

    public static class DockerParams {
        public final String label;
        public final String dir;
        public final Map<Integer, Integer> portMap;

        public DockerParams(String label, String dir, Map<Integer, Integer> portMap) {
            this.label = label;
            this.dir = dir;
            this.portMap = portMap;
        }
    }


    // --- STATIC FIELDS --- //

    private static final String[] DB_ENV = new String[]{"MYSQL_ROOT_PASSWORD=uzabase"};
    private static final String LOAD_FLAG = "MySQL init process done. Ready for start up.";

    public static final int COMMON_PORT = 3361;
    public static final int FINANCE_PORT = 3362;

    public static final DockerParams COMMON = new DockerParams("sqlbatch-spec-test-common", "/docker/common",
            FluentMap.with(3306, COMMON_PORT));

    public static final DockerParams FINANCE = new DockerParams("sqlbatch-spec-test-finance", "/docker/finance",
            FluentMap.with(3306, FINANCE_PORT));



    // --- STATIC METHODS --- //

    public static void runDocker(DockerParams p) throws IOException, InterruptedException {
        if(!com.uzabase.jharmony.docker.DockerUtil.isRunning(p.label)) {
            com.uzabase.jharmony.docker.DockerUtil.insertLogSeparator();

            URL dockerFile = CommonSteps.class.getResource(p.dir);
            if(dockerFile == null) {
                throw new Error("Docker file could not found at " + p.dir);
            }
            String imageId = com.uzabase.jharmony.docker.DockerUtil.build(dockerFile);
            com.uzabase.jharmony.docker.DockerUtil.run(p.label, imageId, p.portMap, ContainerNetwork.bridge(), DB_ENV);

            com.uzabase.jharmony.docker.DockerUtil.waitForContainerWithLabel(p.label,LOAD_FLAG);
        }
    }


    public static void stopDocker(DockerParams p) throws IOException, InterruptedException {
        if(com.uzabase.jharmony.docker.DockerUtil.isRunning(p.label)) {
            com.uzabase.jharmony.docker.DockerUtil.stop(p.label);
        }
    }


    public static void runAll() throws IOException, InterruptedException {
        runDocker(COMMON);
        runDocker(FINANCE);
    }


    public static void stopAll() throws IOException, InterruptedException {
        stopDocker(COMMON);
        stopDocker(FINANCE);
    }


    public static void main(String[] args) {
        try {
            stopAll();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

}

File: mmk2/src/main/scala/co/mscp/mmk2/scala/DockerUtil.scala
package co.mscp.mmk2.scala

import java.net.URL

import co.mscp.mmk2.docker.ContainerNetwork


case class DockerParams(label: String, dir: URL, flag: String,
  portMap: Map[Int, Int], env: Seq[String])
{
  def this(label: String, dir: URL, flag: String, portMap: Map[Int, Int]) =
    this(label, dir, flag, portMap, Seq())
}


object DockerUtil {
  import collection.JavaConverters._

  def insertLogSeparator(): Unit =
    co.mscp.mmk2.docker.DockerUtil.insertLogSeparator()

  def waitForContainerWithLabel(p: DockerParams): Unit =
    co.mscp.mmk2.docker.DockerUtil.waitForContainerWithLabel(p.label, p.flag)

  def build(p: DockerParams): String =
    co.mscp.mmk2.docker.DockerUtil.build(p.dir)

  def run(id: String, p: DockerParams): String =
    co.mscp.mmk2.docker.DockerUtil.run(
      p.label,
      id,
      p.portMap
        .map(pair => (int2Integer(pair._1), int2Integer(pair._2)))
        .asJava,
      ContainerNetwork.bridge(),
      p.env:_*)

  def isRunning(p: DockerParams): Boolean =
    co.mscp.mmk2.docker.DockerUtil.isRunning(p.label)

  def stop(p: DockerParams): Unit = co.mscp.mmk2.docker.DockerUtil.stop(p.label)
}